File: web-client/src/domain/auth/CurrentSession/Session/FirstName.ts
import Assert from "../../../utils/Assert";
import ValidationException from "./ValidationException";

class FirstName {
    public static readonly MAX_LENGTH = 50;
    public static readonly MIN_LENGTH = 1;

    public readonly value: string;

    public constructor(value: string) {
        const cleanedValue = value.trim();

        Assert.notBlank(value, () => new FirstName.MinLengthException());
        Assert.minLength(
            cleanedValue,
            FirstName.MIN_LENGTH,
            () => new FirstName.MinLengthException()
        );
        Assert.maxLength(
            cleanedValue,
            FirstName.MAX_LENGTH,
            () => new FirstName.MaxLengthException()
        );

        this.value = cleanedValue;
    }

    static readonly ValidationException = ValidationException;

    static readonly MinLengthException = class extends FirstName.ValidationException {
        public constructor() {
            super(
                `First name must be at least ${FirstName.MIN_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MinLength";
        };
    };

    static readonly MaxLengthException = class extends FirstName.ValidationException {
        public constructor() {
            super(
                `First name must be at most ${FirstName.MAX_LENGTH} characters long`
            );
        }

        getCode = (): string => {
            return "MaxLength";
        };
    };
}

export default FirstName;

File: backend/microservices/users-ms/src/main/java/com/azat4dev/booking/users/commands/domain/core/values/user/FirstName.java
package com.azat4dev.booking.users.commands.domain.core.values.user;

import com.azat4dev.booking.shared.domain.DomainException;
import com.azat4dev.booking.shared.utils.Assert;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.io.Serializable;

@Getter
@EqualsAndHashCode
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public final class FirstName implements Serializable {

    public static final int MAX_LENGTH = 255;
    private final String value;

    @Override
    public String toString() {
        return value;
    }

    private static void validate(String value) throws CantBeEmptyException, MaxLengthException {
        Assert.string(value, CantBeEmptyException::new).notNull().notBlank();
        Assert.string(value, MaxLengthException::new).maxLength(MAX_LENGTH);
    }

    public static FirstName checkAndMakeFromString(String value) throws CantBeEmptyException, MaxLengthException {

        Assert.notNull(value, CantBeEmptyException::new);
        final var cleanedValue = value.trim();

        validate(cleanedValue);
        return new FirstName(cleanedValue);
    }

    public static FirstName dangerMakeFromStringWithoutCheck(String value) {

        assert value != null;
        return new FirstName(value);
    }

    public abstract static class ValidationException extends DomainException {
        protected ValidationException(String message) {
            super(message);
        }
    }

    public static final class CantBeEmptyException extends ValidationException {
        public CantBeEmptyException() {
            super("The first name cannot be empty");
        }

        @Override
        public String getCode() {
            return "Empty";
        }
    }

    public static final class MaxLengthException extends ValidationException {
        public MaxLengthException() {
            super("The maximum length of first name is " + MAX_LENGTH);
        }

        @Override
        public String getCode() {
            return "MaxLength";
        }
    }
}
